components/service/glean/src/main/java/mozilla/components/service/glean/storages/GenericScalarStorageEngine.kt

```diff
@@ -241,31 +264,35 @@ internal abstract class GenericScalarStorageEngine<ScalarType> : StorageEngine {
 
         // Record a copy of the metric in all the needed stores.
         @SuppressLint("CommitPrefEdits")
-        val userPrefs: SharedPreferences.Editor? =
-            if (metricData.lifetime == Lifetime.User) userLifetimeStorage.edit() else null
-        metricData.getStorageNames().forEach {
-            val storeData = dataStores[metricData.lifetime.ordinal].getOrPut(it) { mutableMapOf() }
+        val editor: SharedPreferences.Editor? = when (metricData.lifetime) {
+            Lifetime.User -> userLifetimeStorage.edit()
+            Lifetime.Ping -> pingLifetimeStorage.edit()
+            else -> null
+        }
+        metricData.getStorageNames().forEach { store ->
+            val storeData = dataStores[metricData.lifetime.ordinal].getOrPut(store) { mutableMapOf() }
             // We support empty categories for enabling the internal use of metrics
             // when assembling pings in [PingMaker].
             val entryName = metricData.identifier
             val combinedValue = combine(storeData[entryName], value)
             storeData[entryName] = combinedValue
-            // Persist data with "user" lifetime
-            if (metricData.lifetime == Lifetime.User) {
+            // Persist data with "user" or "ping" lifetimes
+            editor?.let {
                 serializeSingleMetric(
-                    userPrefs,
-                    "$it#$entryName",
+                    it,
+                    "$store#$entryName",
                     combinedValue,
                     extraSerializationData
                 )
+                it.apply()
```

Comment:
This will trigger the I/O write on each iteration of the `forEach`. Let's leave this back where it was, so that the flush is only triggered once.